File: app/src/main/java/ch/zuehlke/sbb/reddit/features/overview/adapter/impl/RedditNewsDelegateAdapter.java
package ch.zuehlke.sbb.reddit.features.overview.adapter.impl;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.common.base.Strings;
import com.squareup.picasso.Picasso;

import ch.zuehlke.sbb.reddit.R;
import ch.zuehlke.sbb.reddit.features.detail.RedditNewsDetailActivity;
import ch.zuehlke.sbb.reddit.features.overview.adapter.ViewType;
import ch.zuehlke.sbb.reddit.features.overview.adapter.ViewTypeDelegateAdapter;
import ch.zuehlke.sbb.reddit.models.RedditNewsData;
import ch.zuehlke.sbb.reddit.util.DateUtils;

/**
 * Created by chsc on 12.11.17.
 */

public class RedditNewsDelegateAdapter implements ViewTypeDelegateAdapter{

    private final OnNewsSelectedListener mListener;

    public RedditNewsDelegateAdapter(OnNewsSelectedListener listener){
        mListener = listener;
    }

    public interface OnNewsSelectedListener{
        void onNewsSelected(String url);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent) {
        return new RedditNewsViewHolder(parent,parent.getContext());
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, ViewType item) {
        if(holder instanceof RedditNewsViewHolder){
            ((RedditNewsViewHolder) holder).bind((RedditNewsData)item);
        }
    }


    private class RedditNewsViewHolder extends RecyclerView.ViewHolder{

        private ImageView mThumbnail = itemView.findViewById(R.id.thumbnail);
        private TextView mTitle = itemView.findViewById(R.id.description);
        private TextView mAuthor = itemView.findViewById(R.id.author);
        private TextView mComments = itemView.findViewById(R.id.comments);
        private TextView mTime = itemView.findViewById(R.id.time);

        private Context mContext;

        public RedditNewsViewHolder(ViewGroup parent,Context context) {
            super(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_overview, parent, false));
            mContext = context;
        }

        public void bind(final RedditNewsData data){
            mComments.setText(String.valueOf(data.getNumberOfComments()));
            mAuthor.setText(data.getAuthor());
            mTitle.setText(data.getTitle());
            mTime.setText(DateUtils.toFriendlyTime(data.getCreated()));
            if(Strings.isNullOrEmpty(data.getThumbnailUrl())) {
                Picasso.with(mContext).load(R.drawable.reddit_placeholder).into(mThumbnail);
            }else{
                Picasso.with(mContext).load(data.getThumbnailUrl()).into(mThumbnail);
            }

            super.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    mListener.onNewsSelected(data.getPermaLink());
                }
            });
        }
    }
}

File: app/src/main/java/ch/zuehlke/sbb/reddit/features/overview/adapter/impl/RedditNewsDelegateAdapter.kt
package ch.zuehlke.sbb.reddit.features.overview.adapter.impl

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import com.google.common.base.Strings
import com.squareup.picasso.Picasso

import ch.zuehlke.sbb.reddit.R
import ch.zuehlke.sbb.reddit.features.overview.adapter.ViewType
import ch.zuehlke.sbb.reddit.features.overview.adapter.ViewTypeDelegateAdapter
import ch.zuehlke.sbb.reddit.models.RedditNewsData
import ch.zuehlke.sbb.reddit.util.DateUtils

/**
 * Created by chsc on 12.11.17.
 */

class RedditNewsDelegateAdapter(private val mListener: RedditNewsDelegateAdapter.OnNewsSelectedListener) : ViewTypeDelegateAdapter {

    interface OnNewsSelectedListener {
        fun onNewsSelected(url: String)
    }

    override fun onCreateViewHolder(parent: ViewGroup): RecyclerView.ViewHolder {
        return RedditNewsViewHolder(parent, parent.context)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, item: ViewType) {
        if (holder is RedditNewsViewHolder) {
            holder.bind(item as RedditNewsData)
        }
    }


    private inner class RedditNewsViewHolder(parent: ViewGroup, private val mContext: Context) : RecyclerView.ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_overview, parent, false)) {

        private val mThumbnail = itemView.findViewById<ImageView>(R.id.thumbnail)
        private val mTitle = itemView.findViewById<TextView>(R.id.description)
        private val mAuthor = itemView.findViewById<TextView>(R.id.author)
        private val mComments = itemView.findViewById<TextView>(R.id.comments)
        private val mTime = itemView.findViewById<TextView>(R.id.time)

        fun bind(data: RedditNewsData) {
            mComments.text = data.numberOfComments.toString()
            mAuthor.text = data.author
            mTitle.text = data.title
            mTime.text = DateUtils.friendlyTime(data.created)
            if (Strings.isNullOrEmpty(data.thumbnailUrl)) {
                Picasso.with(mContext).load(R.drawable.reddit_placeholder).into(mThumbnail)
            } else {
                Picasso.with(mContext).load(data.thumbnailUrl).into(mThumbnail)
            }

            super.itemView.setOnClickListener { mListener.onNewsSelected(data.permaLink) }
        }
    }
}
